src/common/bitstream.cpp

```diff
@@ -0,0 +1,392 @@
+    if(numBits == 0) {
+        return;
+    }
+
+    // If the stream position is aligned on a byte boundary and we are writing a quantity of bits divisble by 8, we can use faster byte writing
+    if((streamBitPos & 0x7) == 0 && (numBits & 0x7) == 0) {
+        writeBytes(data, numBits / 8);
+        return;
+    }
+
+    // Reserve space for the number of bits we're going to write
+    size_t remainingBits = getRemainingBits();
+    if(remainingBits < numBits) {
+        buf.resize(BITS_TO_BYTES(buf.size() * 8 + numBits - remainingBits));
+    }
+
+    unsigned char* dstPtr = buf.data();
+
+    size_t bitsToWrite = numBits;
+
+    while(true) {
+        size_t byteOffset = streamBitPos / 8;
+        size_t bitOffset = (streamBitPos & 0x7);
+        size_t bitsLeft = 8 - bitOffset;
+
+        if(bitsLeft >= bitsToWrite) {
+            // We have enough room in the current byte to fit all of the rest of the bits, so write them all from the current source byte
+            // Shift the remaining bits left to close the gap and be flush with the end of the stream
+            size_t bitGap = (bitsLeft - bitsToWrite);
+            dstPtr[byteOffset] |= (*data << bitGap);
+            streamBitPos += bitsToWrite;
+            break;
+        }
+        else {
+            // We don't have enough room for all of the remaining bits, so write as much as we need to from the current source byte
+            // Shift the current byte right to un-overlap the bits and be flush with the end of the stream
+            size_t bitsToWriteFromSrc = std::min(bitsToWrite, (size_t)8);
+            size_t bitsOverlapped = bitsToWriteFromSrc - bitsLeft;
+
+            dstPtr[byteOffset] |= (*data >> bitsOverlapped);
+
+            // Now write the rest of the bits remaining in the current source byte to the next destination byte
+            dstPtr[byteOffset + 1] |= (*data << (8 - bitsOverlapped));
+
+            streamBitPos += bitsToWriteFromSrc;
+            bitsToWrite -= bitsToWriteFromSrc;
+
+            if(bitsToWrite == 0) {
+                break;
+            }
+
+            data++;
+        }
+    }
+}
+
+uint16_t BitStream::readStringLength() {
+    uint16_t strLen = 0;
+    readBits((uint8_t*)&strLen, 8);
+    if((strLen & 0x80) != 0) {
+        strLen &= 0x7F;
+    }
+    else {
+        // NOTE: Swapped byte ordering
+        strLen = (strLen << 8);
+        readBits((uint8_t*)&strLen, 8);
+    }
+
+    return strLen;
+}
+
+void BitStream::read(std::string& str) {
+    uint16_t strLen = readStringLength();
+
+    alignPos();
+
+    str.resize(strLen);
+    readBytes((uint8_t*)str.data(), strLen);
+}
+
+void BitStream::read(std::wstring& str) {
+    uint16_t strLen = readStringLength();
+
+    alignPos();
+
+    str.resize(strLen);
+    readBytes((uint8_t*)str.data(), strLen * 2);
+}
+
+void BitStream::readQuantitizedFloat(float& data, const size_t numBits, const float max, const float min) {
```

Comment:
No need to const-ify parameters that are passed by value like size_t and float here. We definitely want const for pointers and references where appropriate so that we can be sure that it won't modify caller data, but whether the function is allowed to manipulate its own copies of the parameters should be an implementation detail left to the function.